cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -320,6 +320,30 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
       return function_converter(substrait_call);
     }
 
+    case substrait::Expression::kCast: {
+      const auto& cast_exp = expr.cast();
+      ARROW_ASSIGN_OR_RAISE(auto input,
+                            FromProto(cast_exp.input(), ext_set, conversion_options));
+
+      // need to convert from the Substrait type spec to the relevant Arrow thing
+      // do we need a switch/case or can we just get code similar to the below FromProto?
```

Comment:
Using `FromProto` as you are doing is the way to go.  The second half of the FromProto return is nullability.  In Arrow nullability is a field concept and not a type concept.  In Substrait it is a part of the type.  So we have to return two pieces when we convert to Arrow.

Acero has no concept of nullability so for the purposes of consumption we generally ignore it (this is a [documented caveat](https://arrow.apache.org/docs/dev/cpp/streaming_execution.html#types)).

This does raise a bit of a pickle when the user requests EXACT_ROUNDTRIP as we may lose information like this.  For example, one thing we will have to consider when implementing the reverse operation is whether the target type should be marked nullable or not.

A limited but safe approach could be to just pick one (probably nullable) and, if we are in EXACT_ROUNDTRIP, then reject any incoming plan that tries to cast to a non-nullable field.

A more extensible approach might be to only reject a plan if it changes the nullability of the input type.  Then, when converting to proto, we would have to make sure the output nullability of the cast matches the input nullability.  I don't yet have a good enough understanding of what the ToProto path will look like to know if this is feasible or not.

Reply:
@westonpace this is a great analysis on the issue. The exact roundtrip was sort of a recurring discussion when we started wrtiting the ToProtos for relations. Can we guarantee the exactness in roundtripping? It is very tricky when we have to make that choice. Should we put a set of rules where we ignore the plans with this kind of inputs or we do not support exact roundtrip for such cases. I am not sure whether this can be generalised. But my understanding is roundtripping could be limited to such complex cases.

Reply:
Thanks! I went for the simpler solution you suggestion for the moment - rejecting the plan if we try to cast to a non-nullable field when strictness is set to `EXACT_ROUNDTRIP`.